crates/bevy_asset/src/loader_builders.rs

```diff
@@ -74,47 +210,108 @@ impl<'ctx, 'builder> NestedLoader<'ctx, 'builder> {
         self.with_transform(move |meta| meta_transform_settings(meta, &settings))
     }
 
-    /// Specify the output asset type.
+    // convert between `T`s
+
+    /// When [`load`]ing, the loader will attempt to load an asset with the
+    /// given [`TypeId`] and [type name].
+    ///
+    /// [type name]: core::any::type_name
     #[must_use]
-    pub fn with_asset_type<A: Asset>(mut self) -> Self {
-        self.asset_type_id = Some(TypeId::of::<A>());
-        self
+    pub fn erased(
+        self,
+        asset_type_id: TypeId,
+        asset_type_name: &'builder str,
+    ) -> NestedLoader<'ctx, 'builder, Erased, M> {
+        NestedLoader {
+            load_context: self.load_context,
+            meta_transform: self.meta_transform,
+            typing: Erased {
+                asset_type_id,
+                asset_type_name,
+            },
+            mode: self.mode,
+        }
     }
 
-    /// Specify the output asset type.
+    /// When [`load`]ing, you must pass in the asset type as a type parameter
+    /// statically.
+    ///
+    /// If you don't know the type statically (at compile time), consider
+    /// [`erased`] or [`unknown_type`].
+    ///
+    /// [`load`]: Self::load
+    /// [`erased`]: Self::erased
+    /// [`unknown_type`]: Self::unknown_type
     #[must_use]
-    pub fn with_asset_type_id(mut self, asset_type_id: TypeId) -> Self {
-        self.asset_type_id = Some(asset_type_id);
-        self
+    pub fn typed(self) -> NestedLoader<'ctx, 'builder, Typed, M> {
+        NestedLoader {
+            load_context: self.load_context,
+            meta_transform: self.meta_transform,
+            typing: Typed { _priv: () },
+            mode: self.mode,
+        }
     }
 
-    /// Load assets directly, rather than creating handles.
+    /// When [`load`]ing, we will infer what type of asset to load from
+    /// metadata.
+    ///
+    /// [`load`]: Self::load
     #[must_use]
-    pub fn direct<'c>(self) -> DirectNestedLoader<'ctx, 'builder, 'c> {
-        DirectNestedLoader {
-            base: self,
-            reader: None,
+    pub fn of_unknown_type(self) -> NestedLoader<'ctx, 'builder, UnknownType, M> {
```

Comment:
I think I'd prefer `unknown_type` rather than `of_unknown_type`? Makes it match `erased` and `typed` more closely, idk.

Reply:
Makes sense to me, I'll rename to `unknown_type`

Reply:
Renamed to `with_unknown_type` to keep it in line with:
- `with_settings`
- `with_[static|dynamic]_type`